File: services/ai/app/geo.py
# services/ai/app/geo.py
"""Geodesy helpers matching `packages/shared/src/utils/geo.ts` (WGS-84 mean radius)."""

from __future__ import annotations

import math
from typing import Sequence

import numpy as np

EARTH_RADIUS_M = 6_371_000.0

# (lon, lat) rings as GeoJSON: exterior first, holes after.
PolygonRings = Sequence[Sequence[tuple[float, float]]]


def to_radians(deg: float) -> float:
    return deg * math.pi / 180.0


def to_degrees(rad: float) -> float:
    return rad * 180.0 / math.pi


def haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Great-circle distance in metres. Same formula as the TypeScript helper."""
    d_lat = to_radians(lat2 - lat1)
    d_lon = to_radians(lon2 - lon1)
    a = (
        math.sin(d_lat / 2.0) ** 2
        + math.cos(to_radians(lat1))
        * math.cos(to_radians(lat2))
        * math.sin(d_lon / 2.0) ** 2
    )
    a = min(1.0, max(0.0, a))
    return 2.0 * EARTH_RADIUS_M * math.atan2(math.sqrt(a), math.sqrt(1.0 - a))


def bearing(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Initial bearing degrees clockwise from north in [0, 360)."""
    phi1 = to_radians(lat1)
    phi2 = to_radians(lat2)
    d_lon = to_radians(lon2 - lon1)
    y = math.sin(d_lon) * math.cos(phi2)
    x = math.cos(phi1) * math.sin(phi2) - math.sin(phi1) * math.cos(phi2) * math.cos(
        d_lon
    )
    return (to_degrees(math.atan2(y, x)) + 360.0) % 360.0


def pairwise_haversine_m(lats: np.ndarray, lons: np.ndarray) -> np.ndarray:
    """NxN great-circle distance matrix in metres."""
    lat1 = np.radians(lats)[:, None]
    lat2 = np.radians(lats)[None, :]
    dlat = lat2 - lat1
    dlon = np.radians(lons)[None, :] - np.radians(lons)[:, None]
    a = np.sin(dlat / 2.0) ** 2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2.0) ** 2
    a = np.clip(a, 0.0, 1.0)
    return 2.0 * EARTH_RADIUS_M * np.arctan2(np.sqrt(a), np.sqrt(1.0 - a))


def _local_xy(
    lat: float, lon: float, lat0: float, lon0: float
) -> tuple[float, float]:
    x = EARTH_RADIUS_M * to_radians(lon - lon0) * math.cos(to_radians(lat0))
    y = EARTH_RADIUS_M * to_radians(lat - lat0)
    return x, y


def point_to_segment_m(
    lat: float,
    lon: float,
    lat_a: float,
    lon_a: float,
    lat_b: float,
    lon_b: float,
) -> float:
    """Geodesic-ish distance from a point to a segment via local equirectangular projection."""
    ax, ay = _local_xy(lat_a, lon_a, lat, lon)
    bx, by = _local_xy(lat_b, lon_b, lat, lon)
    px, py = 0.0, 0.0
    dx, dy = bx - ax, by - ay
    denom = dx * dx + dy * dy
    if denom <= 1e-12:
        return haversine(lat, lon, lat_a, lon_a)
    t = ((px - ax) * dx + (py - ay) * dy) / denom
    t = min(1.0, max(0.0, t))
    cx = ax + t * dx
    cy = ay + t * dy
    # Convert closest XY back to lat/lon around the query point.
    lat_c = lat + to_degrees(cy / EARTH_RADIUS_M)
    cos_lat = math.cos(to_radians(lat))
    if abs(cos_lat) < 1e-12:
        lon_c = lon
    else:
        lon_c = lon + to_degrees(cx / (EARTH_RADIUS_M * cos_lat))
    return haversine(lat, lon, lat_c, lon_c)


def point_to_linestring_m(
    lat: float, lon: float, coordinates: Sequence[tuple[float, float]]
) -> float:
    """Min distance in metres from a point to a LineString of (lon, lat) positions."""
    if len(coordinates) == 0:
        return 0.0
    if len(coordinates) == 1:
        only = coordinates[0]
        return haversine(lat, lon, only[1], only[0])
    best = float("inf")
    for i in range(len(coordinates) - 1):
        a = coordinates[i]
        b = coordinates[i + 1]
        d = point_to_segment_m(lat, lon, a[1], a[0], b[1], b[0])
        if d < best:
            best = d
    return best if math.isfinite(best) else 0.0


def _ray_intersects(
    lon: float, lat: float, ring: Sequence[tuple[float, float]]
) -> bool:
    """Even-odd ray cast. Ring vertices are (lon, lat)."""
    inside = False
    n = len(ring)
    if n < 3:
        return False
    j = n - 1
    for i in range(n):
        xi, yi = ring[i][0], ring[i][1]
        xj, yj = ring[j][0], ring[j][1]
        intersects = ((yi > lat) != (yj > lat)) and (
            lon < (xj - xi) * (lat - yi) / ((yj - yi) + 1e-18) + xi
        )
        if intersects:
            inside = not inside
        j = i
    return inside


def point_in_polygon(lat: float, lon: float, rings: PolygonRings) -> bool:
    """GeoJSON polygon: first ring exterior, subsequent rings holes."""
    if not rings:
        return False
    exterior = rings[0]
    if not _ray_intersects(lon, lat, exterior):
        return False
    for hole in rings[1:]:
        if _ray_intersects(lon, lat, hole):
            return False
    return True

File: packages/shared/src/utils/geo.ts
// packages/shared/src/utils/geo.ts
/** WGS-84 mean Earth radius used by the haversine helpers. */
export const EARTH_RADIUS_M = 6_371_000

export const COORDINATE_DECIMALS = 7

export type LonLat = { lat: number; lon: number }

export type GeoJsonPosition = [number, number] | [number, number, number]

export type GeoJsonPoint = { type: "Point"; coordinates: GeoJsonPosition }
export type GeoJsonLineString = {
  type: "LineString"
  coordinates: GeoJsonPosition[]
}
export type GeoJsonPolygon = {
  type: "Polygon"
  coordinates: GeoJsonPosition[][]
}
export type GeoJsonGeometry = GeoJsonPoint | GeoJsonLineString | GeoJsonPolygon

export type BoundingBox = {
  minLon: number
  minLat: number
  maxLon: number
  maxLat: number
}

const COORD_FACTOR = 10 ** COORDINATE_DECIMALS

export function roundCoordinate(value: number): number {
  return Math.round(value * COORD_FACTOR) / COORD_FACTOR
}
export const roundCoord = roundCoordinate

function toRadians(deg: number): number {
  return (deg * Math.PI) / 180
}

function toDegrees(rad: number): number {
  return (rad * 180) / Math.PI
}

/** Great-circle distance in metres. */
export function haversine(a: LonLat, b: LonLat): number {
  const dLat = toRadians(b.lat - a.lat)
  const dLon = toRadians(b.lon - a.lon)
  const lat1 = toRadians(a.lat)
  const lat2 = toRadians(b.lat)
  const sinDLat = Math.sin(dLat / 2)
  const sinDLon = Math.sin(dLon / 2)
  const h =
    sinDLat * sinDLat + Math.cos(lat1) * Math.cos(lat2) * sinDLon * sinDLon
  return 2 * EARTH_RADIUS_M * Math.atan2(Math.sqrt(h), Math.sqrt(1 - h))
}

/** Initial bearing from `from` to `to`, degrees clockwise from north in [0, 360). */
export function bearing(from: LonLat, to: LonLat): number {
  const lat1 = toRadians(from.lat)
  const lat2 = toRadians(to.lat)
  const dLon = toRadians(to.lon - from.lon)
  const y = Math.sin(dLon) * Math.cos(lat2)
  const x =
    Math.cos(lat1) * Math.sin(lat2) -
    Math.sin(lat1) * Math.cos(lat2) * Math.cos(dLon)
  return (toDegrees(Math.atan2(y, x)) + 360) % 360
}

function isLonLat(value: GeoJsonGeometry | LonLat[] | LonLat): value is LonLat {
  return !Array.isArray(value) && "lat" in value && "lon" in value && !("coordinates" in value)
}

function collectPositions(input: GeoJsonGeometry | LonLat[] | LonLat): LonLat[] {
  if (Array.isArray(input)) return input
  if (isLonLat(input)) return [input]
  switch (input.type) {
    case "Point":
      return [positionFromTuple(input.coordinates)]
    case "LineString":
      return input.coordinates.map(positionFromTuple)
    case "Polygon":
      return input.coordinates.flat().map(positionFromTuple)
    default: {
      const _exhaustive: never = input
      throw new TypeError(`unsupported geometry: ${JSON.stringify(_exhaustive)}`)
    }
  }
}

function positionFromTuple(pos: GeoJsonPosition): LonLat {
  const lon = pos[0]
  const lat = pos[1]
  if (lon === undefined || lat === undefined) {
    throw new TypeError("invalid GeoJSON position")
  }
  return { lon, lat }
}

function bboxFromPoints(points: LonLat[]): BoundingBox {
  if (points.length === 0) {
    throw new TypeError("bbox requires at least one point")
  }
  let minLon = Infinity
  let minLat = Infinity
  let maxLon = -Infinity
  let maxLat = -Infinity
  for (const p of points) {
    if (p.lon < minLon) minLon = p.lon
    if (p.lat < minLat) minLat = p.lat
    if (p.lon > maxLon) maxLon = p.lon
    if (p.lat > maxLat) maxLat = p.lat
  }
  return { minLon, minLat, maxLon, maxLat }
}

/** Axis-aligned bounding box. Pass `radiusM` to pad a point (or the geometry centroid envelope). */
export function bbox(
  input: GeoJsonGeometry | LonLat[] | LonLat,
  radiusM = 0,
): BoundingBox {
  const points = collectPositions(input)
  const box = bboxFromPoints(points)
  if (radiusM <= 0) return box

  const midLat = (box.minLat + box.maxLat) / 2
  const dLat = toDegrees(radiusM / EARTH_RADIUS_M)
  const cosLat = Math.cos(toRadians(midLat))
  const dLon =
    Math.abs(cosLat) < 1e-12
      ? 180
      : toDegrees(radiusM / (EARTH_RADIUS_M * cosLat))
  return {
    minLon: Math.max(-180, box.minLon - dLon),
    minLat: Math.max(-90, box.minLat - dLat),
    maxLon: Math.min(180, box.maxLon + dLon),
    maxLat: Math.min(90, box.maxLat + dLat),
  }
}

function formatOrdinate(n: number): string {
  return String(roundCoordinate(n))
}

function pairToWkt(pos: GeoJsonPosition): string {
  const extra = pos[2] === undefined ? "" : ` ${formatOrdinate(pos[2])}`
  return `${formatOrdinate(pos[0])} ${formatOrdinate(pos[1])}${extra}`
}

function closeRing(ring: GeoJsonPosition[]): GeoJsonPosition[] {
  const first = ring[0]
  const last = ring[ring.length - 1]
  if (
    first &&
    last &&
    first[0] === last[0] &&
    first[1] === last[1] &&
    first[2] === last[2]
  ) {
    return ring
  }
  if (!first) return ring
  return [...ring, first]
}

export function geojsonToWkt(geometry: GeoJsonGeometry): string {
  switch (geometry.type) {
    case "Point":
      return `POINT(${pairToWkt(geometry.coordinates)})`
    case "LineString":
      return `LINESTRING(${geometry.coordinates.map(pairToWkt).join(", ")})`
    case "Polygon":
      return `POLYGON(${geometry.coordinates
        .map((ring) => `(${closeRing(ring).map(pairToWkt).join(", ")})`)
        .join(", ")})`
  }
}

function parseOrdinates(token: string): GeoJsonPosition {
  const parts = token.trim().split(/\s+/).map(Number)
  const lon = parts[0]
  const lat = parts[1]
  if (lon === undefined || lat === undefined || parts.some((n) => Number.isNaN(n))) {
    throw new TypeError(`invalid WKT ordinates: ${token}`)
  }
  const alt = parts[2]
  return alt === undefined ? [lon, lat] : [lon, lat, alt]
}

function splitPairs(body: string): GeoJsonPosition[] {
  return body
    .split(",")
    .map((s) => s.trim())
    .filter((s) => s.length > 0)
    .map(parseOrdinates)
}

function unwrapParens(body: string): string {
  const trimmed = body.trim()
  if (trimmed.startsWith("(") && trimmed.endsWith(")")) {
    return trimmed.slice(1, -1).trim()
  }
  return trimmed
}

/** Destination point after travelling `distanceM` along `bearingDeg` (0 = north). */
export function destination(from: LonLat, distanceM: number, bearingDeg: number): LonLat {
  const angular = distanceM / EARTH_RADIUS_M
  const brng = toRadians(bearingDeg)
  const lat1 = toRadians(from.lat)
  const lon1 = toRadians(from.lon)
  const sinLat1 = Math.sin(lat1)
  const cosLat1 = Math.cos(lat1)
  const sinAng = Math.sin(angular)
  const cosAng = Math.cos(angular)
  const lat2 = Math.asin(sinLat1 * cosAng + cosLat1 * sinAng * Math.cos(brng))
  const lon2 =
    lon1 +
    Math.atan2(
      Math.sin(brng) * sinAng * cosLat1,
      cosAng - sinLat1 * Math.sin(lat2),
    )
  return {
    lat: roundCoordinate(toDegrees(lat2)),
    lon: roundCoordinate(((toDegrees(lon2) + 540) % 360) - 180),
  }
}

/** Approximate circle as a closed GeoJSON polygon (lon, lat). */
export function circlePolygon(
  center: LonLat,
  radiusM: number,
  steps = 24,
): GeoJsonPolygon {
  const count = Math.max(8, Math.round(steps))
  const ring: GeoJsonPosition[] = []
  for (let i = 0; i < count; i += 1) {
    const point = destination(center, radiusM, (i / count) * 360)
    ring.push([point.lon, point.lat])
  }
  const first = ring[0]
  if (first) ring.push([...first])
  return { type: "Polygon", coordinates: [ring] }
}

/** Parse POINT / LINESTRING / POLYGON WKT (optional `SRID=4326;` prefix). */
export function wktToGeojson(wkt: string): GeoJsonGeometry {
  const raw = wkt.trim().replace(/^SRID=\d+\s*;\s*/i, "")
  const match = /^(POINT|LINESTRING|POLYGON)\s*([\s\S]*)$/i.exec(raw)
  if (!match) {
    throw new TypeError(`unsupported WKT: ${wkt}`)
  }
  const kind = match[1]?.toUpperCase()
  const body = match[2]
  if (!kind || body === undefined) {
    throw new TypeError(`unsupported WKT: ${wkt}`)
  }

  if (kind === "POINT") {
    return { type: "Point", coordinates: parseOrdinates(unwrapParens(body)) }
  }
  if (kind === "LINESTRING") {
    return { type: "LineString", coordinates: splitPairs(unwrapParens(body)) }
  }

  const polygonBody = unwrapParens(body)
  const rings: GeoJsonPosition[][] = []
  const ringRe = /\(([^()]*)\)/g
  let ringMatch: RegExpExecArray | null
  while ((ringMatch = ringRe.exec(polygonBody)) !== null) {
    const coords = ringMatch[1]
    if (coords === undefined) continue
    rings.push(splitPairs(coords))
  }
  if (rings.length === 0) {
    throw new TypeError(`invalid POLYGON WKT: ${wkt}`)
  }
  return { type: "Polygon", coordinates: rings }
}
